src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -175,56 +177,50 @@ async ValueTask<(Solution? solution, T result)> TryFastGetSolutionAndRunAsync()
             // See if anyone else is computing this solution for this checksum.  If so, just piggy-back on that.  No
             // need for us to force the same computation to happen ourselves.
             var currentSolution = this.CurrentSolution;
-            var lazySolution = await GetLazySolutionAndIncrementRefCountAsync().ConfigureAwait(false);
 
-            try
-            {
-                // Actually get the solution, computing it ourselves, or getting the result that another caller was computing.
-                var newSolution = await lazySolution.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            // We use a reference-counted solution that implements IAsyncDisposable. The computation of 'newSolution'
+            // uses eager cancellation, but the asynchronous disposable applies lazy cancellation to the final task that
+            // causes cancellation to propagate to the backing lazy operation.
+            var refCountedLazySolution = await GetLazySolutionAsync().ConfigureAwait(false);
+            await using var configuredAsyncDisposable = refCountedLazySolution.ConfigureAwait(false);
 
-                // We may have just done a lot of work to determine the up to date primary branch solution.  See if we
-                // can move the workspace forward to that solution snapshot.
-                if (fromPrimaryBranch)
-                    (newSolution, _) = await this.TryUpdateWorkspaceCurrentSolutionAsync(workspaceVersion, newSolution, cancellationToken).ConfigureAwait(false);
+            // Actually get the solution, computing it ourselves, or getting the result that another caller was computing.
+            var newSolution = await refCountedLazySolution.Target.Task.WithCancellation(cancellationToken).ConfigureAwait(false);
```

Comment:
ok.  would def like a walk through here.  this allows teh caller to cancel asap.  and then the idea is that the disposable will ensure we dispose the lazy solution (Afaict).  however, it's unclear what actually ensures the lazy solution ends up finishing computation and doesn't continue running even once all the callers have returned back out.